R/pipe_consistency_linter.R

```diff
@@ -0,0 +1,83 @@
+#' Pipe consistency linter
+#'
+#' Check that pipe operators are used consistently by file, or optionally
+#' specify one valid pipe operator.
+#'
+#' @param pipe Which pipe operator is valid (either `"%>%"` or `"|>"`). By default
+#' (`"auto"`), the linter has no preference but will check that each file uses
+#' only one type of pipe operator.
+#'
+#' @examples
+#' # will produce lints
+#' lint(
+#'   text = "1:3 |> mean() %>% as.character()",
+#'   linters = pipe_consistency_linter()
+#' )
+#'
+#' lint(
+#'   text = "1:3 %>% mean() %>% as.character()",
+#'   linters = pipe_consistency_linter("|>")
+#' )
+#'
+#' # okay
+#' lint(
+#'   text = "1:3 %>% mean() %>% as.character()",
+#'   linters = pipe_consistency_linter()
+#' )
+#'
+#' lint(
+#'  text = "1:3 |> mean() |> as.character()",
+#'  linters = pipe_consistency_linter()
+#' )
+#' @evalRd rd_tags("pipe_consistency_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+pipe_consistency_linter <- function(pipe = c("auto", "%>%", "|>")) {
+  pipe <- match.arg(pipe)
+
+  xpath_magrittr <- "//SPECIAL[text() = '%>%']"
+  xpath_native <- "//PIPE"
+
+  Linter(function(source_expression) {
+    if (!is_lint_level(source_expression, "file")) {
+      return(list())
+    }
+
+    xml <- source_expression$full_xml_parsed_content
+
+    match_magrittr <- xml2::xml_find_all(xml, xpath_magrittr)
```

Comment:
We've imported `xml_find_all()` into the namespace now, so you no longer need to ns-qualify.